File: 2023/src/day_1.rs
pub fn go(input: String)
{
    let digits = ["1", "2", "3", "4", "5", "6", "7", "8", "9"];
    let sum1 : u32 = input.lines().map(|l| get_cal_val(l, digits.iter())).sum();
    println!("{sum1}");

    let words = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
    let sum2 : u32 = input.lines().map(|l| get_cal_val(l, digits.iter().chain(words.iter()))).sum();
    println!("{sum2}");

}

fn get_cal_val<'a>(line: &str, patterns: impl Iterator<Item = &'a &'static str>) -> u32
{
    let mut matches: Vec<(usize, &str)> = patterns
        .map(|x| line.match_indices(x))
        .flatten()
        .collect();
    matches.sort_by_key(|(i, _)| *i);

    digit_from_string(matches[0].1) * 10 + digit_from_string(matches[matches.len() - 1].1)
}

fn digit_from_string(txt: &str) -> u32
{
    match (txt, txt.parse())
    {
        (_, Ok(n)) => n,
        ("one", _) => 1,
        ("two", _) => 2,
        ("three", _) => 3,
        ("four", _) => 4,
        ("five", _) => 5,
        ("six", _) => 6,
        ("seven", _) => 7,
        ("eight", _) => 8,
        ("nine", _) => 9,
        _ => panic!()
    }
}

File: 2023/Day1.cs
namespace AoC
{
    public class Day1
    {
        public static void Go()
        {
            string[] input = File.ReadAllLines("1.txt");

            string[] digits = ["1", "2", "3", "4", "5", "6", "7", "8", "9"];
            int sum1 = input.Select(l => GetCalVal(l, digits)).Sum();
            Console.WriteLine(sum1);

            string[] words = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
            int sum2 = input.Select(l => GetCalVal(l, digits.Union(words))).Sum();
            Console.WriteLine(sum2);
        }

        static int GetCalVal(string line, IEnumerable<string> patterns)
        {
            string first = patterns.Select(p => (p, i: line.IndexOf(p))).Where(x => x.i != -1).MinBy(x => x.i).p;
            string last = patterns.Select(p => (p, i: line.LastIndexOf(p))).Where(x => x.i != -1).MaxBy(x => x.i).p;
            
            return NumFromString(first) * 10 + NumFromString(last);
        }

        static int NumFromString(string s)
        {
            if (int.TryParse(s, out int n))
            {
                return n;
            }
            else return s switch
            {
                "one" => 1,
                "two" => 2,
                "three" => 3,
                "four" => 4,
                "five" => 5,
                "six" => 6,
                "seven" => 7,
                "eight" => 8,
                "nine" => 9,
                _ => throw new FormatException()
            };
        }
    }
}